server/qa/service.ts

```diff
@@ -0,0 +1,70 @@
+import { Knex } from "knex";
+
+module.exports = class Questions {
+  db: Knex
+
+  constructor(db) {
+    this.db = db;
+  }
+
+  async getAllQuestions(product_id: number, page: number, count: number) {
+    const knex = this.db;
+    const allQuestions = await knex
+      .select('questions.question_id', 'questions.question_body', 'questions.question_date', 'questions.asker_name', 'questions.asker_email', 'questions.question_helpfulness', 'questions.reported')
+      .from('questions')
+      .where({ id_products: product_id })
+      .limit(count)
+      .offset(count * page - count);
+
+    const findAnswersForQuestions = allQuestions.map(async (question) => {
+      const relatedAnswers = await knex
+        .select('answers.answer_id as id', 'answers.body', 'answers.date', 'answers.answerer_name', 'answers.helpfulness', 'answers_photos.url as photos')
+        .from('answers')
+        .where({ id_questions: question.question_id })
+        .leftJoin('answers_photos', 'answers_photos.answer_id', 'answers.answer_id');
+
+      // formatting here to adhere to atelier's nested API data structure
+      const formatted = relatedAnswers.reduce((acc, answer) => (
+        { ...acc, [answer.id]: answer }
+      ), {});
+      return { ...question, answers: formatted };
+    });
+
+    async function questionsAndAnswersResultsWrapper() {
+      const resolvedQuestionsAndAnswers = await Promise.all(findAnswersForQuestions);
+      const resultsWrapper = {
+        product_id,
+        results: resolvedQuestionsAndAnswers,
+      };
+      return resultsWrapper;
+    }
+
+    const wrappedQuestionsAndAnswers = questionsAndAnswersResultsWrapper();
```

Comment:
Is there any discernable benefit to using a separate function to create the wrapper around the data? If not, initializing new functions will just cost you CPU cycles

Reply:
yes, the promise.all needs to be resolved by either using a .then block to chain and extract the result, or use an async function with await to assign it to a variable. in this case using an async function would look cleaner, and have a clearer separation of logic for just the final "wrapping" of the results for readability.

I also believe while it does take up memory, if it can provide clearer code that the performance cost should be negligible.